### ISSUE TITLE
Cache warnings

### ISSUE THREAD

**Issue by rightfold** [purescript/purescript#2308]
**Steps to reproduce**

1. Write a module `M` that causes a warning to be emitted.
2. Compile module `M`.
3. Write a module `N` that imports module `M`.
4. Compile module `N`.

**Expected result**
The warning is emitted in step 4.

**Actual result**
The warning is not emitted, because the module `M` was not being recompiled in step 4.

---

Often I want to fix errors in module `N`, but then the warnings of module `M` aren't emitted anymore, and then I forget to fix those too.

**Comment by natefaubion** [comment]
`psa` has an option to persist warnings https://github.com/natefaubion/purescript-psa#persisting-warnings It can be a little fidgety though if you happen to run compile cycles without psa :/ It would be nice if the compiler wrote warnings to the output directory for a module.

**Comment by rvion** [comment]
> It would be nice if the compiler wrote warnings to the output directory for a module.

I like this idea 👍

**Comment by rightfold** [comment]
Maybe a flag to turn this on per module so that pulp can disable it for bower_components

**Comment by paf31** [comment]
I use a single-file workflow with `pscid`, so I don't tend to notice this as much. I'd just like to note that you could build this sort of workflow by sending `stderr` to a file and building one module at a time (using a `Makefile` or the like). Obviously, that's not always ideal, but I'd like to avoid building support for every workflow into the compiler.

**Comment by rvion** [comment]
>> It would be nice if the compiler wrote warnings to the output directory for a module.

> I like this idea 👍

Second thought: I don't know if it makes sense to write warnings, but if so, let's also write errors

**Comment by natefaubion** [comment]
I don't see how persisting warnings is "building support for every workflow into the compiler". Psc is the single source of truth for warnings, and writing them as part of the module artifacts (like it does with all the other meta info) seems like the least it could do to support other tools. I don't think it needs to get any other options.

**Comment by paf31** [comment]
Sure, I think that's a fine option, but I don't know if the compiler itself should _do_ anything with the warnings, like show them on every rebuild.

**Comment by natefaubion** [comment]
I don't quite see what writing errors gives you. The artifacts didn't even build in that case. To see the errors again, just compile again (which isn't the case with warnings).

**Comment by natefaubion** [comment]
Yeah, I don't think it's necessary for the compiler to do anything with them. Just write them for reference.

**Comment by paf31** [comment]
Presumably we'd save them in the JSON format?

**Comment by rvion** [comment]
> I don't quite see what writing errors gives you. The artifacts didn't even build in that case. To see the errors again, just compile again (which isn't the case with warnings).

I don't say it's a good idea, but I can image several use-cases: 
  - test suites related workflows
  - ability to commit errors-to-be-fixed in some WIP commits
  - ability to cache errors for ide consumptions, (so when you re-open a file, you can see errors without having to wait)
  - etc.

**Comment by natefaubion** [comment]
> Presumably we'd save them in the JSON format?

Yeah, just as is, maybe as `warnings.json` or something? It would be nice to be able to glob them.
